src/main/java/com/faforever/client/game/GenerateMapController.java

```diff
@@ -0,0 +1,246 @@
+      preferencesService.storeInBackground();
+    }));
+  }
+
+  private byte getSliderValue(Slider slider, CheckBox checkBox) {
+    if (checkBox.isSelected()) {
+      return (byte) new Random().nextInt(127);
+    }
+    return (byte) slider.getValue();
+  }
+
+  protected byte[] getOptionArray() {
+    byte spawnCount = spawnCountSpinner.getValue().byteValue();
+    byte landDensity = (byte) (Byte.MAX_VALUE - getSliderValue(waterSlider, waterRandom));
+    byte plateauDensity = getSliderValue(plateauSlider, plateauRandom);
+    byte mountainDensity = getSliderValue(mountainSlider, mountainRandom);
+    byte rampDensity = getSliderValue(rampSlider, rampRandom);
+    return new byte[]{spawnCount, landDensity, plateauDensity, mountainDensity, rampDensity};
+  }
+
+  public void onCloseButtonClicked() {
+    onCloseButtonClickedListener.run();
+  }
+
+  public void onGenerateMapButtonClicked() {
+    onGenerateMap();
+  }
+
+  public void onGenerateMap() {
+    try {
+      if (!previousMapName.getText().isEmpty()) {
+        mapGeneratorService.generateMap(previousMapName.getText()).thenAccept(mapName -> Platform.runLater(() -> {
+          createGameController.initMapSelection();
+          createGameController.mapListView.getItems().stream()
+              .filter(mapBean -> mapBean.getFolderName().equalsIgnoreCase(mapName))
+              .findAny().ifPresent(mapBean -> {
+            createGameController.mapListView.getSelectionModel().select(mapBean);
+            createGameController.mapListView.scrollTo(mapBean);
+            createGameController.setSelectedMap(mapBean);
+          });
+        }));
+      } else {
+        byte[] optionArray = getOptionArray();
+        mapGeneratorService.generateMap(optionArray).thenAccept(mapName -> Platform.runLater(() -> {
+          createGameController.initMapSelection();
+          createGameController.mapListView.getItems().stream()
+              .filter(mapBean -> mapBean.getFolderName().equalsIgnoreCase(mapName))
+              .findAny().ifPresent(mapBean -> {
+            createGameController.mapListView.getSelectionModel().select(mapBean);
+            createGameController.mapListView.scrollTo(mapBean);
+            createGameController.setSelectedMap(mapBean);
+          });
+        }));
+      }
+    } catch (IllegalArgumentException e) {
+      String msg = e.getMessage();
+      if (msg.startsWith("New Version")) {
+        notificationService.addImmediateErrorNotification(e, "mapGenerator.tooNewVersion");
+        log.error("Map generation failed", e);
+      }
+      if (msg.startsWith("Old Version")) {
+        notificationService.addImmediateErrorNotification(e, "mapGenerator.tooOldVersion");
+        log.error("Map generation failed", e);
+      }
+    } catch (Exception e) {
+      notificationService.addImmediateErrorNotification(e, "mapGenerator.generationFailed");
+      log.error("Map generation failed", e);
+    }
```

Comment:
I doubt that is is ever called. You probably need to do this in .excpetionally() on the Completable Future. https://www.logicbig.com/tutorials/core-java-tutorial/java-multi-threading/completion-stages-exception-handling.html

Reply:
I tested it and .exceptionally doesnt work because it is not a completable future that throws the exception rather it is the generatemap function which when it throws the exception never hands back a completable future which is the reason for the try catch but if there is another way it could be structured to use the .exceptionally let me know

Reply:
Nevermind I got it and changed the map generator service to return a completable future that throws an exception

Reply:
Done and let me know if you meant something different

Reply:
Well ok maybe I was wrong then. I just expected it to be thrown inside the CompletableFuture

Reply:
It also should be because otherwise u are doing stuff on the ui thread that u should not do. Lemme see what u did...